### DISCUSSION TITLE
Data List: Data source default value

### DISCUSSION THREAD

**Discussion by bjarnef** [leekelleher/umbraco-contentment#91] [Idea]
### Is your feature request related to a problem? Please describe.

The Data List configuration using .NET Enumeration works great for e.g. Text Size and Text Position in settings of a Block List editor block. However by default in use the first enum value as default value even though no value has been selected in the list editor.

### Describe the solution you'd like

It would be great if we could configurate e.g Checkbox List, Dropdown List or Radio Button List selecting one or more default options.

**Comment by leekelleher** [comment]
Hi @bjarnef, thanks for raising this.

According to the Microsoft docs, the default value of an Enum is `0`...

> The default value of an enumeration type `E` is the value produced by expression `(E)0`, even if zero doesn't have the corresponding enum member. <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum>

This means that it will either use whichever enum member is set to `0`, or the first one. So it'll always have a default value.

If you have control over the Enum, then a workaround would be to set whichever you'd like the default to have a value of `0`.

**Update:** Found a StackOverflow answer with further explanation: <https://stackoverflow.com/a/4967673/12787>

---

I have thought about ways to set a default value, but it started to get complex with the disconnect between the data source and list editor.  Needs more thought and prototyping. 😃

**Comment by bjarnef** [comment]
@leekelleher yeah, I know. However I think in some cases if might make sense to have a default option in the UI, while keeping the default enum values - especially if using the enum in configuration of multiple data type instances.

E.g. a WeekDay or .NET `DayOfWeek` enum. You might use the default values to make it easier to work with `Calendar` and `DateTime` objects in .NET, but for some reason the client might prefer Friday to be selected as default in Radiobutton List.
https://docs.microsoft.com/en-us/dotnet/api/system.dayofweek?view=net-5.0

Furthermore if used in Checkbox List it would be might be useful to make Monday - Friday selected by default, while Saturday and Sunday are unselected.

**Comment by leekelleher** [comment]
@bjarnef, thank you - I do like the idea of setting a default value, _(whether this is for Enum data source or any other),_ but I struggle with where (conceptually) the user would configure that?

e.g. is it the responsibility of the data source? or the list editor? _(but then I'd designed the list editor to be agnostic of data source)_

I could argue that this could be done using Content Templates (blueprints), but then BlockList/NestedContent doesn't support those, (although [Contentment's Content Blocks](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/content-blocks.md) does).

Or a last resort (as a developer), programmatically set a default value using a [`EditorModelEventManager.SendingContentModel` hook](https://our.umbraco.com/documentation/reference/events/EditorModel-Events/#usage)?

I guess this comes down to whether we need a solution for a specific problem, or it's a more conceptual question, _should Data List be able to define a default value?_

Hopefully, I'm not sounding like I'm trying to shut down any discussions... I'm totally happy to hear any proposed ideas. 🙏

**Reply by hfloyd** [reply]
@leekelleher In answer to:

> I do like the idea of setting a default value, _(whether this is for Enum data source or any other),_ but I struggle with where (conceptually) the user would configure that?
> 
> e.g. is it the responsibility of the data source? or the list editor? _(but then I'd designed the list editor to be agnostic of data source)_

It seems that which value would be considered the "default" should be set on the DataList itself... exactly how to do that is a bit trickier... 

1. You could change the underlying Data source object model to something like this: 
- Prop: DataItems = IEnumerable of current list of items (Name, Value, Icon, Description on each)
- Prop: DefaultValue =  Which Item is default. (How this gets set would vary based on the specific source, but it could be configured in the Custom Datasource code... For the basic "type in a list via the UI" datasources, it can be a simple string storing the "default" Value - since those should be unique across items). Empty string would indicate "no specified default".
I can think of several possible issues with this approach... it isn't perfect, but it insures that only one value is set as default.

2. Another option is to add another property to the DataItem model - a Boolean for "IsDefault" - the main problem is managing User error (setting more than one item with "IsDefault=true"). Of course, when actually determining which IS the default, if more than one are found, the code could just take the first one where "IsDefault=true". The site implementer/developer would realize the problem pretty quickly if they didn't get the default they were expecting.

Options 1 and 2 both keep the idea of the "default" items WITH the items.

Then... On the list editor, any editor for which it makes sense could have a simple boolean option - "Use Default Value" if set to true, then that would be the initial value of the data on a new node, for instance. There could always be another "Allow blank" boolean, so if that is also true, the Content Editor would be able to "clear" the default value (thus storing a null value).

When the list editor is setting the default value on new/null content, if the Data source includes an option marked as the default, then that is used, and if no specific option is marked as the default, then it can just use the first option in the list automatically. 


So, to summarize, wherever the data is entered/stored, each datalist has a default value (either explicitly set, or the first item in the list). And applicable list editors have an option to use it or not.

At the very least, I'd like for the appropriate built-in list editors (radio, button, etc.) to allow for a default value to be set on new content. (I think right now only the drop-down list has the option.) If we can't get to the point where a specific default can be specified, I'd be satisfied with the first option being assumed default.

**Reply by hfloyd** [reply]
Actually... considering @bjarnef 's comment about checkbox lists.... I think I like the idea of the data item-level "IsDefault" (Option 2, above) a bit more. Technically,. multiple of the Data items could then be marked as "default", and if the list editor is a multi-select type, ALL those defaults could be "checked"/selected.

My angular skillz are lacking... but I can envision the general back-end logic to handle these things...

**Comment by bjarnef** [comment]
@leekelleher I imagine this is something to configurate in the list editor (or maybe overwrite the value from data source).

With a enum data source and a checkbox list, I guess it isn't possible to set multiple default checked values?

In a specific use-case I am using enum data source + radiobutton list for some block lidt configuration settings.

**Text Position**
Left, Center, Right
In this case when no value is selected "Left" is default, which is fine in this case.

**Text Width**
Small, Medium, Large
In this case when no value is selected "Small" is default, but actually I would like to have "Large" as default as it represent full width :)

With checkbox list I can imagine use-cases where it would be useful to set multiple default values (maybe common used values or something which should be enabled 90% of the times, so it is just easier to uncheck the other 10% of the times).

**Comment by Matthew-Wise** [comment]
Ran into this issue today as well, when using drop down list in my case the first option been default was fine however the dropdown was not pre populated so I had to add some code using `EditorModelEventManager.SendingContentModel`.

I think the control of default value should sit with the editor, this would allow the data source to be reused with different default values based on its use case. This would also allow the checkbox list to have multiple selected, as its in control.

**Comment by leekelleher** [comment]
I'm not convinced that the List Editor is the appropriate place to configure a default value - since it knows nothing about the data source, (and making it aware of the data source would introduce much complexity).

Potential idea could be to make use of the proposed Preview panel, see PR #53. Which would display a rendering of the configured Data List editor.

![Data List Preview idea](https://user-images.githubusercontent.com/209066/99831422-f014aa00-2b56-11eb-9bf3-0b3c9b3d0726.png)

But I'm not 100% on this, conceptually in Umbraco we do not (generally) define default values at Data Type level - which is why Content Templates exist. (I'm aware this isn't ideal for sub-content editors, a la BlockList, NC, DTGE, et al) 🤔

**Comment by Matthew-Wise** [comment]
I can see you're point about defaults and content templates.

Maybe it's more of a documentation task to highlight setting a default value using code? (by linking out to our or a more complex example)

Not sure if you could set a default value by data type definition easily using  `EditorModelEventManager.SendingContentModel` though, it would have to be by document type / property definition.

**Comment by leekelleher** [comment]
Quick thought about an default value for the Enum specifically... I wonder if the object-type should be Nullable Enum instead? 🤔

**Comment by Matthew-Wise** [comment]
> Quick thought about an default value for the Enum specifically... I wonder if the object-type should be Nullable Enum instead? 🤔

If allow empty or generally?

**Comment by leekelleher** [comment]
General, (as the Data Source knows nothing about the List Editor's config).  It could be configurable at the Enum Data Source level... but then I start to feel it's death by config options, which is what I nearly fell into with issue #42 (many granular options vs. a single sensible one)

**Comment by leekelleher** [comment]
Moved this topic to a discussion. I'm still unsure how to best approach a "default value" feature for Data List.
Conflicted with the concept, in terms of how Umbraco handles default values for properties. Totally open for further discussion and brainstorming.

**Comment by greystate** [comment]
Just came here after searching for how to set a default value :)

I'd made DataList with **User-defined list** of "Smaller", "Normal" & "Bigger" as the Data source, and the **Buttons** List editor - my rendering code treats no value/null/empty and "Normal" as the same - but I'd of course like to have the "Normal" button pressed as default too. I actually thought I could do it in the Preview as mentioned earlier, and when that didn't seem to work, I had a look at the JSON to see if maybe there was a secret "default" key available 😁

I can see why it's tricky, having swappable sources and editors though... 🤔

**Reply by leekelleher** [reply]
It's a tricky one. Where should default values be set? (in terms of general Umbraco) 🤷 
I had looked at adding this feature, but didn't fully sit comfortably with me, (and felt like I was over-complicating it), so settled on either using Content Templates (Blueprints) or hooking into the `EditorModelEventManager` (as mentioned above).

**Comment by leekelleher** [comment]
Closing this discussion, as since Umbraco 15, you can define your own [`propertyValuePreset` extension](https://docs.umbraco.com/umbraco-cms/extend-your-project/backoffice-extensions/extending-overview/extension-types/property-value-preset) to programmatically set the default value of a property-editor.
